### PR TITLE
mcp: fix nil params crash vulnerability (#186)

### LINKED ISSUE #186: Able to crash a server from improper handling of nil params
**Describe the bug**
When using the `go-sdk` library, when a no parameters were parsed to the sdk with certain methods. It caused the mcp server to crash. This looks like it could be a vector for effective denial of service. Example of affected methods are: `resources/read` and `prompts/get` but other methods are most likely affected as well.

We suspect this is the problematic function: https://github.com/modelcontextprotocol/go-sdk/blob/main/mcp/server.go#L358

**To Reproduce**
**1.** Start the MCP service.

**2.** Send an `initialiize`
```{"jsonrpc":"2.0","id":"1\"{{3*3}}","method":"initialize","protocolVersion":"2025-06-11","params":{"logging":"true"}}```

**3.** Send a `resources/read` with no parameters.
```{"jsonrpc":"2.0","id":"1", "method": "resources/read"}```

**4.** Observe that you get a `panic: runtime error: invalid memory address or nil pointer dereference`

**Expected behavior**
Proper handling of params should be done to avoid a panic runtime error and return an appropriate error message.

**Logs**

```
mcp-issue-nil-params % go run hello.go 
{"jsonrpc":"2.0","id":"1\"{{3*3}}","method":"initialize","protocolVersion":"2025-06-11","params":{"logging":"true"}}
read: {"jsonrpc":"2.0","id":"1\"{{3*3}}","method":"initialize","params":{"logging":"true"}}
{"jsonrpc":"2.0","id":"1\"{{3*3}}","result":{"capabilities":{"completions":{},"logging":{},"prompts":{"listChanged":true},"resources":{"listChanged":true},"tools":{"listChanged":true}},"protocolVersion":"2025-06-18","serverInfo":{"name":"greeter","version":""}}}
write: {"jsonrpc":"2.0","id":"1\"{{3*3}}","result":{"capabilities":{"completions":{},"logging":{},"prompts":{"listChanged":true},"resources":{"listChanged":true},"tools":{"listChanged":true}},"protocolVersion":"2025-06-18","serverInfo":{"name":"greeter","version":""}}}
{"jsonrpc":"2.0","id":"1", "method": "resources/read"}
read: {"jsonrpc":"2.0","id":"1","method":"resources/read"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x8 pc=0x1006269a8]
```

### PR DESCRIPTION
**Pull request by davidleitw**
Handle nil RawMessage and explicit JSON "null" in unmarshalParams to prevent server crashes. When JSON-RPC requests are sent without parameters or with null parameters, ensure orZero receives a valid struct instead of nil.

**Fixes #186** - DoS vulnerability where missing params caused server panic

**Changes:**
- Add nullJSON constant for efficient null comparison
- Handle nil RawMessage by creating empty parameter struct
- Handle explicit "null" JSON using bytes.Equal
- Integrate with upstream methodFlags system for proper param validation
- Add comprehensive test coverage for nil parameter scenarios

**Behavior:**
- Methods with `missingParamsOK` flag: gracefully handle nil/null params
- Methods without flag: return proper error instead of crashing
- Maintains backward compatibility with existing error handling

**Security Impact:**
- Eliminates DoS attack vector via malformed JSON-RPC requests
- Server continues running instead of crashing on bad input

### CODE REVIEW

File: mcp/shared.go [hunk 1]
```diff
@@ -33,6 +34,9 @@ var supportedProtocolVersions = []string{
 	"2024-11-05",
 }
 
+// Precomputed null JSON for efficient comparison (issue #186 fix)
```
Comment (jba) [inline mcp/shared.go]: Remove this comment

File: mcp/shared.go [hunk 2]
```diff
@@ -217,21 +221,34 @@ func newMethodInfo[S Session, P paramsPtr[T], R Result, T any](d typedMethodHand
...
+				if flags&missingParamsOK == 0 {
+					return nil, fmt.Errorf("%w: missing required \"params\"", jsonrpc2.ErrInvalidRequest)
+				}
+				return orZero[Params](new(T)), nil
+			}
+
+			// Handle explicit JSON "null" (params set to null)
+			if bytes.Equal(m, nullJSON) {
+				if flags&missingParamsOK == 0 {
+					return nil, fmt.Errorf("%w: missing required \"params\"", jsonrpc2.ErrInvalidRequest)
```
Comment (jba) [inline mcp/shared.go]: Remove quotation marks around Params

**Review by jba — CHANGES_REQUESTED** [review]

**Comment by davidleitw** [comment]
@jba Thanks for the review! I've actually updated this PR based on the discussion in #186 - it now only contains test improvements and no longer includes the fix logic (since #210 already resolved the core issue). The code you're commenting on has been removed.

 The PR now just adds complementary unit test coverage for the methodFlags system.

**Review by findleyr — DISMISSED** [review]
Thanks!

**Comment by findleyr** [comment]
It looks like this PR does not format cleanly. Could you please run `go fmt`?

**Comment by davidleitw** [comment]
ok~ let me fix it
